airflow/jobs/scheduler_job.py

```diff
@@ -1417,6 +1420,42 @@ def _clean_tis_without_dagrun(self, session):
                     raise
             guard.commit()
 
+    @provide_session
+    def _missing_dag_file_cleanup(self, session: Session = None):
+        """Fails task instances and DagRuns of DAGs that no longer exist in the dag folder/DB"""
+        states_to_check = State.unfinished - frozenset([State.NONE, State.SHUTDOWN])
+        tis = session.query(TI).filter(TI.state.in_(states_to_check)).all()
+        missing_dags = {}
+        dag_runs = set()
+
+        for ti in tis:
+            if ti.dag_id in missing_dags:
+                ti.set_state(State.FAILED, session=session)
+                continue
+            # Dag no longer in dagbag?
+            if not self.dagbag.has_dag(ti.dag_id, session=session):
+                ti.set_state(State.FAILED, session=session)
+                dag_runs.add(ti.dag_run)
+                missing_dags[ti.dag_id] = [ti]
```

Comment:
Does we need to store `ti` here?

`missing_dags` can be just a set

```python
missing_dags = set()
```

and then this line can be:

```suggestion
                missing_dags.add(ti.dag_id)
```